node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/util/ConfigServerHttpRequestExecutor.java

```diff
@@ -97,6 +97,8 @@ public static ConfigServerHttpRequestExecutor create(
                     NODE_ADMIN_LOGGER.warning("Failed to communicate with " + configServer + ", will try next: " + e.getMessage());
                 }
                 lastException = e;
+                // reshuffle when failure, to avoid always trying in the same order. TODO: Should rather keep track of when and which server failed
+                Collections.shuffle(configServerHosts);
```

Comment:
How does this work? The list it's iterating over with the for-loop is shuffled - does that mean the for-loop is unaffected, or affected by the shuffling? Seems clearer to loop using an index.

Also, this class needs to be thread-safe - using shuffle on List doesn't seem thread-safe.

Alternatively:
 - If no servers are found to work, keep the order of the list.
 - Otherwise, if the working server is not the first: Make a new list with the working list as the first element, and then all other elements in the same order. Then update the (volatile) configServerHosts.

Reply:
Yeah, thanks, this is not good. We also have similar code in RetryingJaxRsStrategy, which does not do this right either, maybe there should be some class that both of those could use. Will close this PR.